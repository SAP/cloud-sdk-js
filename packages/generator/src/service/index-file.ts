import {
  ExportDeclarationStructure,
  SourceFileStructure,
  StructureKind
} from 'ts-morph';
import { VdmServiceMetadata } from '../vdm-types';
import { hasEntities } from '../generator-utils';

/**
 * @internal
 */
export function indexFile(service: VdmServiceMetadata): SourceFileStructure {
  return {
    kind: StructureKind.SourceFile,
    statements: [
      ...service.entities.map(entity => exportStatement(entity.className)),
      ...service.entities.map(entity =>
        exportStatement(`${entity.className}RequestBuilder`)
      ),
      ...service.complexTypes.map(complexType =>
        exportStatement(complexType.typeName)
      ),
      ...service.enumTypes.map(enumType => exportStatement(enumType.typeName)),
      ...(service.functionImports && service.functionImports.length
        ? [exportStatement('function-imports')]
        : []),
      ...(service.actionImports && service.actionImports.length
        ? [exportStatement('action-imports')]
        : []),
      ...(hasEntities(service) ? [exportStatement('BatchRequest')] : []),
      exportStatement('service')
    ]
  };
}

function exportStatement(moduleName: string): ExportDeclarationStructure {
  return {
    kind: StructureKind.ExportDeclaration,
    moduleSpecifier: `./${moduleName}`
  };
}
