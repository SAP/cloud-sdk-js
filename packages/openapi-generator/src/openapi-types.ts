import { OpenAPIV3 } from 'openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of a parsed OpenApi Service.
 */
export interface OpenApiDocument {
  apiName: string;
  operations: OpenApiOperation[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of an operation.
 */
export interface OpenApiOperation extends OpenAPIV3.OperationObject {
  operationName: string;
  method: string;
  pattern: string;
  requestBody?: OpenApiRequestBody;
  parameters: OpenApiParameter[];
}

const supportedMethods = {
  get: 'get',
  put: 'put',
  post: 'post',
  patch: 'patch',
  delete: 'delete',
  head: 'head',
  options: 'options'
} as const;

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Methods supported by OpenApi and SAP Cloud SDK.
 */
export type Method = typeof supportedMethods[keyof typeof supportedMethods];

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Get supported methods.
 * @returns Methods supported by OpenApi and SAP Cloud SDK.
 */
export function methods(): Method[] {
  return Object.values(supportedMethods);
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenatiation of a parameter for both queries and path parameters.
 */
export interface OpenApiParameter extends OpenAPIV3.ParameterObject {
  type: string;
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenation of the request body.
 */
export interface OpenApiRequestBody extends OpenAPIV3.RequestBodyObject {
  parameterName: string;
  parameterType: string;
}
