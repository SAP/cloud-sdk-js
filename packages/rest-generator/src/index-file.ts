import { SourceFileStructure, StructureKind } from 'ts-morph';
export function indexFile(): SourceFileStructure {
  return {
    kind: StructureKind.SourceFile,
    statements: [
      {
        kind: StructureKind.ExportDeclaration,
        moduleSpecifier: './open-api/model'
      }
    ]
  };
}
