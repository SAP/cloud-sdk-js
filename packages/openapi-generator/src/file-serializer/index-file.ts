import { codeBlock, kebabCase } from '@sap-cloud-sdk/util';
import { OpenApiDocument } from '../openapi-types';

/**
 * Serialize an index file for the root directory of the generated service.
 * @param openApiDocument - The parsed document to serialize into an index file.
 * @returns The serialized index file contents.
 * @internal
 */
export function apiIndexFile(openApiDocument: OpenApiDocument): string {
  const files = [
    ...openApiDocument.apis.map(api => api.name),
    ...(openApiDocument.schemas.length ? ['schema'] : [])
  ];
  return codeBlock`
    ${exportAllFiles(files.map(fileName => kebabCase(fileName)))}
  `;
}

/**
 * Serialize an index file for the schema directory of the generated service.
 * @param openApiDocument - The parsed document to serialize into an index file for the schema directory.
 * @returns The serialized index file contents.
 * @internal
 */
export function schemaIndexFile(openApiDocument: OpenApiDocument): string {
  return exportAllFiles(openApiDocument.schemas.map(schema => schema.fileName));
}

function exportAllFiles(fileNames: string[]): string {
  return codeBlock`${fileNames
    .map(fileName => exportAll(fileName))
    .join('\n')}`;
}

function exportAll(file: string) {
  return `export * from './${file}';`;
}
