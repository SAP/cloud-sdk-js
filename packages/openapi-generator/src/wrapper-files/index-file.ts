import { codeBlock } from '@sap-cloud-sdk/util';
import voca from 'voca';
import { OpenApiDocument } from '../openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Generate the index file to expose the SAP Cloud SDK API and models.
 * @param openApiDocument Parsed service.
 * @returns The index file contents.
 */
export function indexFile(openApiDocument: OpenApiDocument): string {
  return codeBlock`
export * from './openapi/model';
${getApiFilesForIndex(openApiDocument)
  .map(api => exportAll(api))
  .join('\n')}
`;
}

function exportAll(file: string) {
  return `export * from './${file}';`;
}

function getApiFilesForIndex(openApiDocument: OpenApiDocument): string[] {
  return openApiDocument.tags.map(tag => buildApiFileNameUsedByIndexFile(tag));
}

function buildApiFileNameUsedByIndexFile(apiName: string) {
  return `${voca.kebabCase(apiName + 'Api')}`;
}
