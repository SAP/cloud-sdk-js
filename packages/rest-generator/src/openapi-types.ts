import { $Refs } from '@apidevtools/swagger-parser';
import { OpenAPIV3 } from 'openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of a parsed OpenApi Service.
 */
export interface OpenApiDocument {
  apiName: string;
  serviceDirName: string;
  refs: $Refs;
  operations: OpenApiOperation[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of an operation.
 */
export interface OpenApiOperation extends OpenAPIV3.OperationObject {
  method: string;
  pattern: string;
  requestBody?: OpenApiRequestBody;
  parameters: OpenApiParameter[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Methods supported by OpenApi and SAP Cloud SDK.
 */
export enum Method {
  get = 'get',
  put = 'put',
  post = 'post',
  patch = 'patch',
  delete = 'delete',
  head = 'head',
  options = 'options'
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenatiation of a parameter for both queries and path parameters.
 */
export interface OpenApiParameter extends OpenAPIV3.ParameterObject {
  type: string;
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenation of the request body.
 */
export interface OpenApiRequestBody extends OpenAPIV3.RequestBodyObject {
  name: string;
  type: string;
}
