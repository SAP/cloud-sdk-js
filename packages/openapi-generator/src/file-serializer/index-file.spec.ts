import { OpenApiDocument } from '../openapi-types';
import { apiIndexFile, schemaIndexFile } from './index-file';

it('apiIndexFile serializes the api index file with referenced schemas', () => {
  const document = {
    apis: [{ name: 'TestApi' }, { name: 'DefaultApi' }],
    schemas: [{}]
  } as OpenApiDocument;
  expect(apiIndexFile(document)).toMatchInlineSnapshot(`
    "    export * from './test-api';
        export * from './default-api';
        export * from './schema';"
  `);
});

it('apiIndexFile serializes the api index file without referenced schemas', () => {
  const document = {
    apis: [{ name: 'TestApi' }, { name: 'DefaultApi' }],
    schemas: []
  } as unknown as OpenApiDocument;
  expect(apiIndexFile(document)).toMatchInlineSnapshot(`
    "    export * from './test-api';
        export * from './default-api';"
  `);
});

it('schemaIndexFile serializes the schema index file for schemas in a document', () => {
  const document = {
    apis: [],
    schemas: [
      { schemaName: 'MySchema1', fileName: 'my-schema-1' },
      { schemaName: 'MySchema2', fileName: 'some-other-name' }
    ]
  } as unknown as OpenApiDocument;
  expect(schemaIndexFile(document)).toMatchInlineSnapshot(`
    "export * from './my-schema-1';
    export * from './some-other-name';"
  `);
});
