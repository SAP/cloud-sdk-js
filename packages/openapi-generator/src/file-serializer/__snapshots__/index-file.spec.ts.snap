// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index-file apiIndexFile serializes the api index file following the esm format 1`] = `
"    export * from './test-api.js';
    export * from './default-api.js';
    export * from './schema/index.js';"
`;

exports[`index-file apiIndexFile serializes the api index file with referenced schemas 1`] = `
"    export * from './test-api';
    export * from './default-api';
    export * from './schema';"
`;

exports[`index-file apiIndexFile serializes the api index file without referenced schemas 1`] = `
"    export * from './test-api';
    export * from './default-api';"
`;

exports[`index-file schemaIndexFile serializes the schema index file for schemas following the ESM format 1`] = `
"export * from './my-schema-1.js';
export * from './some-other-name.js';"
`;

exports[`index-file schemaIndexFile serializes the schema index file for schemas in a document 1`] = `
"export * from './my-schema-1';
export * from './some-other-name';"
`;
